### PR TITLE
[ALLUXIO-3089]change to ConcurrentHashSet to void endless-loop

### PR DESCRIPTION
**Pull request by Mitisky**
https://alluxio.atlassian.net/browse/ALLUXIO-3089

### CODE REVIEW

**Comment by apc999** [comment]
@Mitisky  can you elaborate on the issue? e.g., what issues we see here? how would concurrent hashmap helps here?

**Comment by Mitisky** [comment]
@apc999 OK.The case I met is when the MEM is full and keep writing or reading data from the MEM in same worker, the worker is possible to crash when it evicts block.I printed the stack of JVM and found a endless-loop that occur in the `HashMap$TreeNode.root` method. The `HashSet` is not a thread safe class.It's easy to get a linked loop when `HashMap`  resize the capacity  same time.`ConcurrentHashSet` can void the problem I think.

the stack info:
pool-4-thread-254" #26014 prio=5 os_prio=0 tid=0x00007f2ab0ae4800 nid=0x2d6e runnable [0x00007f28b1e10000]
   java.lang.Thread.State: RUNNABLE
	at java.util.HashMap$TreeNode.root(HashMap.java:1816)
	at java.util.HashMap$TreeNode.getTreeNode(HashMap.java:1881)
	at java.util.HashMap.getNode(HashMap.java:575)
	at java.util.HashMap.containsKey(HashMap.java:595)
	at java.util.HashSet.contains(HashSet.java:203)
	at alluxio.worker.block.meta.StorageDirView.isMarkedToMoveOut(StorageDirView.java:166)
	at alluxio.worker.block.BlockMetadataManagerView.isBlockMarked(BlockMetadataManagerView.java:137)
	at alluxio.worker.block.BlockMetadataManagerView.isBlockEvictable(BlockMetadataManagerView.java:125)
	at alluxio.worker.block.BlockMetadataManagerView.getBlockMeta(BlockMetadataManagerView.java:218)
	at alluxio.worker.block.evictor.AbstractEvictor.cascadingEvict(AbstractEvictor.java:137)
	at alluxio.worker.block.evictor.AbstractEvictor.freeSpaceWithView(AbstractEvictor.java:176)
	at alluxio.worker.block.TieredBlockStore.freeSpaceInternal(TieredBlockStore.java:591)
	at alluxio.worker.block.TieredBlockStore.createBlockMeta(TieredBlockStore.java:194)
	at alluxio.worker.block.DefaultBlockWorker.createBlock(DefaultBlockWorker.java:289)
	at alluxio.worker.block.BlockWorkerClientServiceHandler$7.call(BlockWorkerClientServiceHandler.java:212)
	at alluxio.worker.block.BlockWorkerClientServiceHandler$7.call(BlockWorkerClientServiceHandler.java:208)
	at alluxio.RpcUtils.call(RpcUtils.java:62)
	at alluxio.worker.block.BlockWorkerClientServiceHandler.requestBlockLocation(BlockWorkerClientServiceHandler.java:208)
	at alluxio.thrift.BlockWorkerClientService$Processor$requestBlockLocation.getResult(BlockWorkerClientService.java:1027)
	at alluxio.thrift.BlockWorkerClientService$Processor$requestBlockLocation.getResult(BlockWorkerClientService.java:1011)
	at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
	at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
	at org.apache.thrift.TMultiplexedProcessor.process(TMultiplexedProcessor.java:123)
	at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:286)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

**Comment by yupeng9** [comment]
@apc999 I think this issue is possible to happen.

Looking at the trace, `TieredBlockStore` stores only one instance of `Evictor` that is created in the constructor.  In method `freeSpace`, `mEvictor.freeSpaceWithView` is called which triggered the concurrency issue. 

`freeSpace` is called for every heartbeat via `reserveSpace`. That means if freeSpace takes longer than 1 heartbeat, then this issue might be triggered. I am not sure if this is possible during really heavy load, but I think it's more defensive to make `freeSpace` thread-safe

**Comment by apc999** [comment]
@yupeng9  I think from the trace stack, there is possibly a race. I am not convinced though having the set concurrent will fix this issue. I think we should find out the root cause of this race. And what you suggest might be the possibility. 

@calvinjia  any input here since you worked recently on the eviction side?

**Comment by calvinjia** [comment]
@apc999 I did not encounter this issue in my async eviction testing. Aren't we protected by the metadata lock to prevent concurrent access to the views?

**Comment by yupeng9** [comment]
@calvinjia where is the metadata lock you talked about? From the trace, it seems the race comes from `DefaultBlockWorker.createBlock ` -> `TieredBlockStore.createBlock` -> `TieredBlockStore.freeSpaceInternal` ->`AbstractEvictor.freeSpaceWithView`.

Since the view is shared within the process, it can be accessed by any `createBlock` invocations?

**Comment by calvinjia** [comment]
@yupeng9 Yeah it looks like we only take a read lock when in fact we are doing write operations (updating the Evictor's view).

**Comment by yupeng9** [comment]
@calvinjia  I see. For a simple fix, how about we add `synchronized` to `createBlock`? btw, According to ALLUXIO-1624, TieredBlockStore should be thread-safe

**Comment by yupeng9** [comment]
@Mitisky can you try to use `mMetadataWriteLock` instead of `mMetadataReadLock` at `freeSpaceInternal`, in particular:

```
try (LockResource r = new LockResource(mMetadataReadLock)) {
      plan = mEvictor.freeSpaceWithView(availableBytes, location, getUpdatedView());
      // Absent plan means failed to evict enough space.
      if (plan == null) {
        throw new WorkerOutOfSpaceException(ExceptionMessage.NO_EVICTION_PLAN_TO_FREE_SPACE);
      }
    }
```

So that `freeSpaceWithView` is guarded by a write lock.

Can you make this change and see if you can reproduce the endless-loop issue?

**Comment by apc999** [comment]
@Mitisky do you mind to give a shot on the fix suggested by @yupeng9 ?
if it works, we would like to put your patch in v1.7.1.

**Comment by Mitisky** [comment]
@apc999 @yupeng9 It works. I run the same test case several times. Everything is ok.

**Comment by yupeng9** [comment]
@Mitisky do you think your test case can be packaged as a unit test?

File: core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java [hunk 1]
```diff
@@ -676,7 +676,7 @@ private TempBlockMeta createBlockMetaInternal(long sessionId, long blockId,
   private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLocation location)
       throws WorkerOutOfSpaceException, IOException {
     EvictionPlan plan;
-    try (LockResource r = new LockResource(mMetadataReadLock)) {
+    try (LockResource r = new LockResource(mMetadataWriteLock)) {
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java]: can you add a comment here
mentioning that grabbing the write lock due to ALLUXIO-3089?
Ideally this should still use read lock

**Comment by apc999** [comment]
@Mitisky any update? feel free to ask help from our side.

**Comment by Mitisky** [comment]
@apc999 Thank you. I add a unit test about this issue.

**Comment by apc999** [comment]
@Mitisky thanks for the unit test! Without your fix in `TieredBlockStore.java`, will the unit tests fail due to the end-less loop?

**Comment by Mitisky** [comment]
@apc999 yes.If `TieredBlockStore#freeSpaceInternal` is still guaranteed by a read lock,this unit test will fail because the endless-loop causes a timeout exception

**Pull request merged**